### PR TITLE
feature/group-rbac

### PR DESCRIPTION
**Pull request by dovgopoly**

### CODE REVIEW

File: contracts/interfaces/access-control/IRBAC.sol [hunk 1]
```diff
@@ -7,6 +7,12 @@ import "../../libs/data-structures/StringSet.sol";
  *  @notice The RBAC module
  */
 interface IRBAC {
+    enum PermissionStatus {
```
Comment (Arvolear) [inline contracts/interfaces/access-control/IRBAC.sol:10]: Naah, let's just use true and false

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        address who_,
+        string[] memory groupsToRemoveFrom_
+    ) public virtual override onlyPermission(RBAC_RESOURCE, DELETE_PERMISSION) {
+        require(groupsToRemoveFrom_.length > 0, "RBACGroupable: empty groups");
+
+        _removeUserFromGroups(who_, groupsToRemoveFrom_);
+    }
+
+    function grantGroupRoles(
+        string calldata groupTo_,
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:40]: Please use `memory` in the contract. The interface is fine

File: contracts/access-control/RBAC.sol [hunk 3]
```diff
@@ -191,41 +191,8 @@ abstract contract RBAC is IRBAC, Initializable {
...
-                _allowed.contains(ALL_PERMISSION) ||
-                _allowed.contains(permission_)
-            ) {
-                isAllowed_ = true;
-            }
-        }
-
-        return isAllowed_;
+    ) public view virtual override returns (bool) {
+        return _hasPermission(who_, resource_, permission_) == PermissionStatus.AllowsPermission;
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:195]:
```suggestion
        return _hasRolesPermission(getUserRoles(who_), resource_, permission_);
```

File: contracts/access-control/RBAC.sol [hunk 4]
```diff
@@ -296,4 +263,50 @@ abstract contract RBAC is IRBAC, Initializable {
...
     }
+
+    function _hasPermission(
+        address who_,
+        string memory resource_,
+        string memory permission_
+    ) internal view virtual returns (PermissionStatus) {
+        return _getRolesPermissionStatus(getUserRoles(who_), resource_, permission_);
+    }
+
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:274]:
```suggestion
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        _groupRoles[groupFrom_].remove(rolesToRevoke_);
+
+        emit RevokedGroupRoles(groupFrom_, rolesToRevoke_);
+    }
+
+    function _hasPermission(
+        address who_,
+        string memory resource_,
+        string memory permission_
+    ) internal view virtual override returns (PermissionStatus) {
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:98]:
```suggestion
    function hasPermission(
        address who_,
        string memory resource_,
        string memory permission_
    ) public view override returns (bool) {
```

File: contracts/interfaces/access-control/extensions/IRBACGroupable.sol [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+interface IRBACGroupable {
```
Comment (Arvolear) [inline contracts/interfaces/access-control/extensions/IRBACGroupable.sol:4]: Please use `calldata` everywhere

File: contracts/access-control/RBAC.sol [hunk 7]
```diff
@@ -191,41 +191,8 @@ abstract contract RBAC is IRBAC, Initializable {
...
-                _allAllowed.contains(permission_) ||
-                _allowed.contains(ALL_PERMISSION) ||
-                _allowed.contains(permission_)
-            ) {
-                isAllowed_ = true;
-            }
-        }
-
-        return isAllowed_;
+    ) public view virtual override returns (bool) {
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:194]:
```suggestion
    ) public view override returns (bool) {
```

File: contracts/access-control/RBAC.sol [hunk 8]
```diff
@@ -296,4 +263,50 @@ abstract contract RBAC is IRBAC, Initializable {
...
+        string memory permission_
+    ) internal view virtual returns (PermissionStatus) {
+        return _getRolesPermissionStatus(getUserRoles(who_), resource_, permission_);
+    }
+
+    function _getRolesPermissionStatus(
+        string[] memory roles_,
+        string memory resource_,
+        string memory permission_
+    ) internal view returns (PermissionStatus) {
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:279]:
```suggestion
    ) internal view returns (PermissionStatus rolesPermissionStatus_) {
```

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 5 above)
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:98]:
```suggestion
    ) internal view virtual override returns (PermissionStatus userPermissionStatus_) {
```

File: contracts/interfaces/access-control/IRBAC.sol [hunk 9]
```diff
@@ -7,6 +7,12 @@ import "../../libs/data-structures/StringSet.sol";
...
+        DoesNotHavePermission,
+        AllowsPermission,
+        DisallowsPermission
+    }
```
Comment (Arvolear) [inline contracts/interfaces/access-control/IRBAC.sol:14]:
```suggestion
    enum PermissionStatus {
        None,
        Allow,
        Disallow
    }
```

File: contracts/interfaces/access-control/IRBAC.sol (hunk 1 above)
Comment (Arvolear) [inline contracts/interfaces/access-control/IRBAC.sol:10]: Please move this enum to the contract. It is internal and should not be in the interface

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        );
+
+        if (userPermissionStatus_ == PermissionStatus.DisallowsPermission) {
+            return userPermissionStatus_;
+        }
+
+        string[] memory groups_ = getUserGroups(who_);
+
+        for (uint256 i = 0; i < groups_.length; i++) {
+            PermissionStatus rolesPermissionStatus_ = _getRolesPermissionStatus(
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:112]:
```suggestion
            userPermissionStatus_ = _getRolesPermissionStatus(
```

Is this a useless variable?

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+                permission_
+            );
+
+            if (rolesPermissionStatus_ == PermissionStatus.DisallowsPermission) {
+                return rolesPermissionStatus_;
+            }
+
+            if (rolesPermissionStatus_ == PermissionStatus.AllowsPermission) {
+                userPermissionStatus_ = rolesPermissionStatus_;
+            }
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:124]:
```suggestion
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        for (uint256 i = 0; i < groups_.length; i++) {
+            PermissionStatus rolesPermissionStatus_ = _getRolesPermissionStatus(
+                getGroupRoles(groups_[i]),
+                resource_,
+                permission_
+            );
+
+            if (rolesPermissionStatus_ == PermissionStatus.DisallowsPermission) {
+                return rolesPermissionStatus_;
+            }
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:120]:
```suggestion
            if (userPermissionStatus_ == PermissionStatus.DisallowsPermission) {
                return PermissionStatus.Disallows;
            }
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 13]
```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import "../../interfaces/access-control/extensions/IRBACGroupable.sol";
+
+import "../../libs/arrays/SetHelper.sol";
+
+import "../RBAC.sol";
+
+abstract contract RBACGroupable is RBAC, IRBACGroupable {
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:10]:
```suggestion
abstract contract RBACGroupable is IRBACGroupable, RBAC {
```

File: contracts/access-control/RBAC.sol [hunk 14]
```diff
@@ -296,4 +263,50 @@ abstract contract RBAC is IRBAC, Initializable {
...
+
+    function _hasPermission(
+        address who_,
+        string memory resource_,
+        string memory permission_
+    ) internal view virtual returns (PermissionStatus) {
+        return _getRolesPermissionStatus(getUserRoles(who_), resource_, permission_);
+    }
+
+    function _getRolesPermissionStatus(
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:275]:
```suggestion
    function _hasRolesPermission(
```

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 2 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/interfaces/access-control/extensions/IRBACGroupable.sol (hunk 6 above)
Comment (dovgopoly) [reply contracts/interfaces/access-control/extensions/IRBACGroupable.sol:4]: done

File: contracts/access-control/RBAC.sol (hunk 7 above)
Comment (dovgopoly) [reply contracts/access-control/RBAC.sol]: done

File: contracts/access-control/RBAC.sol (hunk 8 above)
Comment (dovgopoly) [reply contracts/access-control/RBAC.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 5 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/interfaces/access-control/IRBAC.sol (hunk 9 above)
Comment (dovgopoly) [reply contracts/interfaces/access-control/IRBAC.sol]: done

File: contracts/interfaces/access-control/IRBAC.sol (hunk 1 above)
Comment (dovgopoly) [reply contracts/interfaces/access-control/IRBAC.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 13 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/access-control/RBAC.sol (hunk 14 above)
Comment (dovgopoly) [reply contracts/access-control/RBAC.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 12 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/access-control/RBAC.sol [hunk 15]
```diff
@@ -296,4 +273,54 @@ abstract contract RBAC is IRBAC, Initializable {
...
+     *  @param resource_ the role resource to search the permission in
+     *  @param permission_ the permission to search
+     *  @return true_ if the role has the permission, false otherwise
+     */
+    function _isAllowed(
+        string memory role_,
+        string memory resource_,
+        string memory permission_
+    ) internal view returns (bool) {
+        mapping(string => StringSet.Set) storage _allAllowedResources = _rolePermissions[role_][
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:289]:
```suggestion
        mapping(string => StringSet.Set) storage _resources = _rolePermissions[role_][
```

File: contracts/access-control/RBAC.sol [hunk 16]
```diff
@@ -296,4 +273,54 @@ abstract contract RBAC is IRBAC, Initializable {
...
+     *  @param resource_ the role resource to search the antipermission in
+     *  @param permission_ the antipermission to search
+     *  @return true_ if the role has the antipermission, false otherwise
+     */
+    function _isDisallowed(
+        string memory role_,
+        string memory resource_,
+        string memory permission_
+    ) internal view returns (bool) {
+        mapping(string => StringSet.Set) storage _allDisallowedResources = _rolePermissions[role_][
```
Comment (Arvolear) [inline contracts/access-control/RBAC.sol:314]:
```suggestion
        mapping(string => StringSet.Set) storage _resources = _rolePermissions[role_][
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+
+import "../../libs/arrays/SetHelper.sol";
+
+import "../RBAC.sol";
+
+/**
+ *  @notice The Role Based Access Control (RBAC) module
+ *
+ *  This contract is an extension for the RBAC contract to provide the ability to organize roles
+ *  into groups and assign groups to users.
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:14]:
```suggestion
 *  into groups and assign users to them.
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+    /**
+     *  @notice The init function
+     */
+    function __RBACGroupable_init() internal onlyInitializing {
+        __RBAC_init();
+    }
+
+    /**
+     *  @notice The function to assign groups to the user
+     *  @param who_ the user to assign groups to
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:32]:
```suggestion
     *  @param who_ the user to be assigned
```

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+     *  @param groupsToAddTo_ the list of groups to be assigned
```
Comment (Arvolear) [inline contracts/access-control/extensions/RBACGroupable.sol:33]:
```suggestion
     *  @param groupsToAddTo_ the list of groups to assign the user to
```

File: contracts/access-control/RBAC.sol (hunk 15 above)
Comment (dovgopoly) [reply contracts/access-control/RBAC.sol]: done

File: contracts/access-control/RBAC.sol (hunk 16 above)
Comment (dovgopoly) [reply contracts/access-control/RBAC.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 19 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 18 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+
+    /**
+     *  @notice The init function
+     */
+    function __RBACGroupable_init() internal onlyInitializing {
+        __RBAC_init();
+    }
+
+    /**
+     *  @notice The function to assign groups to the user
```
[the remark this answers is not in the document]
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

File: contracts/access-control/extensions/RBACGroupable.sol (hunk 17 above)
Comment (dovgopoly) [reply contracts/access-control/extensions/RBACGroupable.sol]: done

**Review by Arvolear — APPROVED** [review]
Let's merge this bad boy

**Pull request merged**